### PR TITLE
Adding better error message when service account token is not found

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref https://github.com/kubernetes/kubernetes/issues/19053

Should help in debugging e2e flake

cc @liggitt

### CODE REVIEW

File: plugin/pkg/admission/serviceaccount/admission.go [hunk 1]
```diff
@@ -284,6 +294,10 @@ func (s *serviceAccount) getReferencedServiceAccountToken(serviceAccount *api.Se
 		}
 	}
 
+	if errIfTokenNotFound {
+		return "", fmt.Errorf("no secret found which is a service acount token for service account  %s/%s, secrets: %v, service account tokens: %v",
```
Comment (liggitt) [inline plugin/pkg/admission/serviceaccount/admission.go]: I don't see how this is any clearer than the error returned on line 350

**Comment by liggitt** [comment]
not sure how these errors help debug... I'd rather keep the existing error as is. If you want to add higher verbosity glog tracing, that might be ok

**Comment by nikhiljindal** [comment]
I want to know that why we reach that error on line 350.
Is it because there are no secrets or because no service account tokens?

**Comment by liggitt** [comment]
service account tokens are secrets

**Comment by liggitt** [comment]
like I said, I'm ok with trace logging

**Comment by liggitt** [comment]
also, in the case of e2e, it's almost certainly the first branch... the controller manager creates a secret containing the token, then adds a reference to that secret to the service account. Checking if the service account has the reference is cheaper, which is why we do that first. In e2e, that's what I would expect to be hitting, but that doesn't really tell us if the secret was created already or not

**Comment by nikhiljindal** [comment]
fwiw, its not the first branch that is failing.
I have got it to fail twice locally and both times the error was `no service account token found for service account`

**Comment by liggitt** [comment]
ok. the admission controller is working off a cache store... it's possible the store hasn't been updated with the token if a pod create attempt is made immediately after the token and service account are updated, but before the watch propagates

**Comment by nikhiljindal** [comment]
In getServiceAccount(), we do a direct lookup, if its not found in cache.
I guess, we should do the same in getServiceAccountTokens()?

Will try that change and  see if  the test still fails

**Comment by liggitt** [comment]
I wouldn't... that has the potential to fan out a lot (1..N tokens per service account)

File: plugin/pkg/admission/serviceaccount/admission.go (hunk 1 above)
Comment (lavalamp) [inline plugin/pkg/admission/serviceaccount/admission.go]: It tells you which particular part of the process went wrong, no?

Comment (liggitt) [inline plugin/pkg/admission/serviceaccount/admission.go]: but it loses all the detail the message on line 350 adds about what you actually need to do

**Comment by liggitt** [comment]
I still think I prefer the existing message... if you don't feel strongly, can we close this?

**Pull request closed without merging**